Implement mechanism for game-specific tweaks and workarounds

dreamer opened the pull request:
Quote from tweak module docstring:

> This module lists game-specific tweaks, that cannot be handled by other means (specifically tweaks, that are outside of dxvk.conf or Mesa scope, such as additional environment variables, renaming files post-installation, changing Wine settings for the specific game, etc.).  The purpose is similar to "Install Scripts" *.vfd files distributed with games, except here tweaks are provided by Proton community and not game developer.

To show usage, I included workarounds for three games:

* STAR WARS™ Jedi Knight - Jedi Academy™, appid 6020
* STAR WARS™ Jedi Knight II - Jedi Outcast™, appid 6030
* Call of Duty® (2003), appid 2620

After these fixes, Jedi Academy could be whitelisted, Jedi Outcast requires additional work (I have no sound and certain id Tech 3 - specific workaround needs to be injected to fix "grey filter" bug). I don't own Call of Duty (I included fix based on [reports on SPCR](https://spcr.netlify.com/app/2620)).

I tested Star Wars titles on two systems: [NVIDIA](https://gist.github.com/dreamer/03179db6f82b01e06ec8fa41faf09f00) and [Mesa/Intel](https://gist.github.com/dreamer/4f3773020ded88d4612e41de4dd67b7e).

pylint reports 10/10 on tweaks.py module

I have additional functionality for Tweaks class in development (importing Core Fonts installed in users' OS, which would help to run multiple titles, including e.g. GTA V). I decided to not include more functionality with this PR, because it needs more testing and I want to learn if Proton maintainers are interested in this mechanism at all.

Some ideas, how this class could be expanded in future:

* switching Wine windows version in appid prefix
* injecting new default settings into games
* providing game-specific defaults for Proton environment variables
* installing additional dependencies missing from Install Script included in windows depot

dreamer committed:
Create tweaks module

This module implements barebones Tweaks class, which will be used for
implementing game-specific tweaks and workarounds, that cannot be
implemented by other means.

So far this class enables injection of environment variables, that were
vetted by Proton developers.

dreamer committed:
Remove executable flag from user_settings file

dreamer committed:
Add env variables for some id Tech 3 games

Fixes following titles:

* Call of Duty® 2003 (2620)
* STAR WARS™ Jedi Knight - Jedi Academy™ (6020)
* STAR WARS™ Jedi Knight II - Jedi Outcast™ (6030)

All of these games were released around 2003 on id Tech 3 engine, which
used statically allocated buffer for querying OpenGL extensions.

Limit number of OpenGL extensions on AMD/Intel:
MESA_EXTENSION_MAX_YEAR=2003

Limit number of OpenGL extensions on proprietary NVIDIA:
__GL_ExtensionStringVersion=17700

dreamer commented:
Oh, maybe I should include a short list of titles, that *could* run better if started with some additional environment variables (compiled based on SPCR reports):

* Battle Brothers
* Battlerite
* Bioshock
* Bioshock 2
* Claire
* LEGO® Star Wars™ - The Complete Saga
* LEGO® Indiana Jones™: The Original Adventures
* Sonic Generations Collection
* Thief (2014)

... and probably many more. When browsing through reports, the most common workarounds I saw included: disabling esync, other OpenGL related flags and PulseAudio related flags.

If Tweaks included functionality installing Core Fonts in game prefix, it would significantly improve the situation (there are many situations where games do not respect font replacements provided by wine and in result either game is unplayable or various game launchers cannot start). Also, Core Fonts installed in prefix solve issue, where Install Script tries to install vcredist libraries and keeps failing.

ihusky02 commented:
Did you can make this tweak for world of tanks blitz? (Ap ID 444200)

nanonyme commented:
Am I misunderstanding or is this hardcoding an nVidia proprietary driver hack most likely done because they implemented core profiles wrong into Proton?

dreamer commented:
> Did you can make this tweak for world of tanks blitz? (Ap ID 444200)

This game worked for me perfectly without any tweaks. What exactly do you have in mind?

nanonyme commented:
Ah, there's also Mesa-specific hacks. https://www.mesa3d.org/envvars.html you might want to document in the change that this is a workaround to too small fixed size buffer for reading gl extensions in the game
Same is most likely the case with the nVidia-specific value

dreamer commented:
> Am I misunderstanding or is this hardcoding an nVidia proprietary driver hack most likely done because they implemented core profiles wrong into Proton?

No, that's not it. There is a bug in the game engine, which was fixed sometime after these games were released.

Place of OpenGL-related tweaks like these would be in OpenGL implementation, so outside of Proton but rather in Mesa or NVIDIA driver. While we theoretically could implement a workaround in Mesa, it would affect only AMD users - it wouldn't work for Intel users, because their drivers do not have per-binary workarounds, and it wouldn't work for NVIDIA users, because if NVIDIA didn't include it so far, they will never do it in their Linux driver.

The way I am proposing in this PR, it will work for everyone using Proton, out of the box.

> (...) you might want to document in the change that this is a workaround to too small fixed size buffer for reading gl extensions in the game. Same is most likely the case with the nVidia-specific value

I already did document it, in [commit message](https://github.com/ValveSoftware/Proton/pull/1423/commits/23f6a28a8b04dc1a4857d93ad2dbbb6770a07852) implementing the change :)

ihusky02 commented:
> > Did you can make this tweak for world of tanks blitz? (Ap ID 444200)
> 
> This game worked for me perfectly without any tweaks. What exactly do you have in mind?

Game shop, tournaments, messages section don't work ( when click on this, game sudenlly close). I think this hapen because net framework and corefonts are not added in proton or in game prefix.

dreamer commented:
> > > Did you can make this tweak for world of tanks blitz? (Ap ID 444200)
> > 
> > This game worked for me perfectly without any tweaks. What exactly do you have in mind?
> 
> Game shop, tournaments, messages section don't work ( when click on this, game sudenlly close). I think this hapen because net framework and corefonts are not added in proton or in game prefix.

Then I suggest installing corefonts using [protontricks](https://github.com/Sirmentio/protontricks) to test if it'll resolve the issue for you: `$ protontricks 444200 corefonts`.

I'll be monitoring #1293 to see if I can include fixes for this particular game in future patches.

Let's keep discussion here focused on general mechanism and not specific games :)

nanonyme commented:
Sorry, missed that somehow. The mechanism does indeed sound sensible

ihusky02 commented:
> > > > Did you can make this tweak for world of tanks blitz? (Ap ID 444200)
> > > 
> > > 
> > > This game worked for me perfectly without any tweaks. What exactly do you have in mind?
> > 
> > 
> > Game shop, tournaments, messages section don't work ( when click on this, game sudenlly close). I think this hapen because net framework and corefonts are not added in proton or in game prefix.
> 
> Then I suggest installing corefonts using [protontricks](https://github.com/Sirmentio/protontricks) to test if it'll resolve the issue for you: `$ protontricks 444200 corefonts`.
> 
> I'll be monitoring #1293 to see if I can include fixes for this particular game in future patches.
> 
> Let's keep discussion here focused on general mechanism and not specific games :)

Tell me how to do this? I need only write this command in terminal?

dreamer commented:
@ihusky02 I left a comment for you in #1293 

@kisak-valve  @aeikum @Nephyrin What do you think about this PR? Should I keep working on this or do you prefer some other solution?

Elkasitu commented on tweaks.py:
```diff
@@ -0,0 +1,71 @@
+}
+
+
+class Tweaks:  # pylint: disable=too-few-public-methods
```
Unless I'm missing something, this is unnecessary, you can just export `TWEAKS_DB` and do `env.update(TWEAKS_DB.get(env.get("SteamAppId")).get("env", {}))`

The pylint error about too few public methods is a hint :wink:

dreamer replied:
Oh yes, I know - as I explained in PR, I am developing additional functionalities to this class, and this warning is already disabled in my local checkout :)

nanonyme replied:
This looks reasonably hard to read in such a packed form though

dreamer replied:
@nanonyme I agree. Putting this line straight into `proton` file would also make it harder to implement unit tests in future. And Proton definitely could use some tests right now...

Elkasitu replied:
```python
game_tweaks = Tweaks(env.get("SteamAppId"))
env.update(game_tweaks.env)
```
to

```python
game_tweaks = TWEAKS_DB.get(env.get("SteamAppId"), {})
env.update(game_tweaks.get("env", {}))
```

is as readable, you could use even more intermediate variables to make it super readable, if you wanted to

dreamer replied:
... unless I want to use `game_tweaks.import_fonts()` in some other place in code ;) But I hear you - maybe it would be better to implement future tweaks as more functions in `proton` file, but I wanted to separate responsibilities early. `proton` script is long and non-pythonic in several ways already.

Elkasitu commented on tweaks.py:
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python
+
+# This is free and unencumbered software released into the public domain.
```
I doubt the compatibility of this license with Proton's

dreamer replied:
I copied it from `filelock.py` distributed with Proton. But I probably should change hashbang to require `python3`.

dreamer replied:
Changed to python3 (I am not going to write code that works in both python2 and python3 at the same time, it's a hassle).

Elkasitu reviewed (commented):
This is a good idea, but perhaps a better implementation would be to fetch the tweaks from a real online db, and users could activate tweaks from the steam client ui (something like "automatically improve game compatibility") but I know that this requires infrastructure, Valve (or someone with enough resources) would need to set up a "wine game tweak db" that can be queried once during game setup/update, in this manner new game tweaks could be used **without updating Proton**, but this is a good start.

Maybe we could use a google document / github repo as a nosql db, or something like that

Just a quick suggestion: Tweaks can be defined in a data format instead of a source code format (python) such as `csv`, `json` or even `xml`, there's no logic in the `.py` file and the aforementioned formats have the advantage of having GUIs for easily performing CRUD operations, this is not the case for python.

dreamer commented:
@Elkasitu 
> would need to set up a "wine game tweak db" that can be queried once during game setup/update, in this manner new game tweaks could be used without updating Proton

I think it might be overkill - dxvk tweaks are already distributed with Proton updates and it might be beneficial to keep this file in sync with wine and dxvk versions (maybe a new version of Wine will turn some workarounds moot?). By keeping this in repo, we create an easy way for the community to engage implementation of tweaks (e.g. PRs can be referenced from Whitelist requests).

Also, tweaks might need to be implemented not only during setup/update but maybe on every start of the game (it's already the case for using Core Fonts installed in OS).

> Maybe we could use a google document / github repo as a nosql db, or something like that

Anything but Google Docs, please :) I was helping with [SPCR](https://spcr.netlify.com/) Google sheet and it, erhm, "didn't work very well" (SPCR almost migrated away from gsheets at this point).

> Just a quick suggestion: Tweaks can be defined in a data format instead of a source code format (python) such as csv, json or even xml, there's no logic in the .py file and the aforementioned formats have the advantage of having GUIs for easily performing CRUD operations, this is not the case for python.

That was my idea (and first implementation) as well, but decided to switch to simple python dict for the following reasons:

* it's simpler, less code, fewer opportunities for error
* it's faster (but I can't imagine this database to grow too big, so it's very tiny advantage)
* in dict we can include python functions as values - this might prove useful
* Only tweaks, that are required for a game to run/not_crash/provide_full_functionality should be included here. Performance tuning and such should be left for users to make on their own. Games should just work, without setting up checkboxes in Steam UI or using a separate tool like `protontricks`. I imagine sometimes a tweak will improve a game for some users and make it worse for others - in such cases, it should be excluded from this file.

ihusky02 commented:
protontricks don`t work so good. I can`t install vcrun 2012 and 2015 because  it`s don`t change wineprefix to 32 bit (for Ap ID 444200) and I have returned error status 5 with vredist_x86.exe ehhhh  I don`t know why it work when I use Wine and install steam client use winetricks but don`t work with native steam for linux and protontricks.

Ahhh and add information that protontricks work when winetricks is copied into /usr/bin/ directory and STEAM_DIR AND PROTONVERSION is set

nanonyme commented:
I do hope it's enough that it's in PATH. If not, can you link to source so I can do PR?

ihusky02 commented:
> I do hope it's enough that it's in PATH. If not, can you link to source so I can do PR?

No I think it`s because steam play or protontricks don`t do something like winetricks (WINEARCH=win32 WINEPREFIX=/home/_youruserdir/.wine winecfg) <=== create 32bit prefix before install important things and game. Otherwise I don`t know where is place steam play wine, because I can`t get into path /Proton 3.7/dist/ because I have error "too many arguments" or file or directiory doesn`t exist.

dreamer commented:
@ihusky02 this thread is for discussing Pull Request, not usage of `winetricks` or `protontricks`. Continue this discussion in #1293, please.

dreamer committed:
Use python3 in tweaks module

dreamer committed:
tweaks: Install Core Fonts if found in OS

When trying to run a game, that depends on corefonts to run correctly,
Proton looks for corefonts installed in OS and hard link or copy
them to expected location in a game prefix.

No additional registry entries are required (at least for titles, that
I tested so far), Wine automatically registers these fonts when a game
starts.

Tweaks module will tries to find fonts using fc-list (for system and
user locations), then it looks into default Wine prefix (~/.wine/),
and finally into /usr/share/fonts/* locations, where corefonts are
likely to reside.

dreamer committed:
tweaks: Add titles requiring Core Fonts

List of titles was obtained by analyzing Steam Play Community Reports.

dreamer commented:
I updated Tweaks with the ability to import corefonts from OS. Quote from commit message:

> When trying to run a game, that depends on corefonts to run correctly, Proton looks for corefonts installed in OS and hard link or copy them to expected location in a game prefix.
>
> No additional registry entries are required (at least for titles, that I tested so far), Wine automatically registers these fonts when a game starts.
>
> Tweaks module tries to find fonts using fc-list (for system and user locations), then it looks into default Wine prefix (~/.wine/), and finally into /usr/share/fonts/* locations, where corefonts are likely to reside.

I scanned through [SPCR reports](https://spcr.netlify.com/raw-reports) and included tweaks for games, that I am quite sure will benefit from this change (tested some of those titles to confirm).

Next step would be to bundle Proton with FLOSS replacements for Windows fonts (the issue is discussed here: #571), at packaging time install them in `dist/share/default_pfx` and register as replacements for their original counterparts (this is especially important for non-western fonts). After that, we could trim `TWEAKS_DB` to include only titles, that depend on original fonts.

@kisak-valve @aeikum I am waiting for your response about this PR - do you like where this is going?

ihusky02 commented:
> I updated Tweaks with the ability to import corefonts from OS. Quote from commit message:
> 
> > When trying to run a game, that depends on corefonts to run correctly, Proton looks for corefonts installed in OS and hard link or copy them to expected location in a game prefix.
> > No additional registry entries are required (at least for titles, that I tested so far), Wine automatically registers these fonts when a game starts.
> > Tweaks module tries to find fonts using fc-list (for system and user locations), then it looks into default Wine prefix (~/.wine/), and finally into /usr/share/fonts/* locations, where corefonts are likely to reside.
> 
> I scanned through [SPCR reports](https://spcr.netlify.com/raw-reports) and included tweaks for games, that I am quite sure will benefit from this change (tested some of those titles to confirm).
> 
> Next step would be to bundle Proton with FLOSS replacements for Windows fonts (the issue is discussed here: #571), at packaging time install them in `dist/share/default_pfx` and register as replacements for their original counterparts (this is especially important for non-western fonts). After that, we could trim `TWEAKS_DB` to include only titles, that depend on original fonts.
> 
> @kisak-valve @aeikum I am waiting for your response about this PR - do you like where this is going?

can you tell me how use this script?

dreamer commented:
> can you tell me how use this script?

The proper way would be to [build Proton](https://github.com/ValveSoftware/Proton#building-for-linux) from [my branch](https://github.com/dreamer/Proton/tree/po/tweaks) and replace the version downloaded automatically by Steam (that's the way I do it). But I tested my changes with WoT:Blitz, it did not resolve crashes in that game (that's why it is not included in `TWEAKS_DB`).

ihusky02 commented:
> > can you tell me how use this script?
> 
> The proper way would be to [build Proton](https://github.com/ValveSoftware/Proton#building-for-linux) from [my branch](https://github.com/dreamer/Proton/tree/po/tweaks) and replace the version downloaded automatically by Steam (that's the way I do it). But I tested my changes with WoT:Blitz, it did not resolve crashes in that game (that's why it is not included in `TWEAKS_DB`).

Talk me more, when I am green user like total noob, so I must do this command on clean install native steam  client (without enable steam play)? But how to replace proton version?

OK, let me know if you find solution my friend, good luck :) and have a nice day

dreamer committed:
proton: Introduce __main__ scope

This is first step towards making proton testable.  Also, comments in
main script functionality formatted according to PEP-8.

dreamer committed:
tweaks: Implement commandline tweaking

dreamer committed:
tweaks: Fix grey screen in Jedi Kinght II

dreamer committed:
tweaks: Implement user override for tweaks

dreamer committed:
proton: pass modified args to dump_dbg_scripts

dreamer committed:
tweaks: Remove pointless copyright

This module is already covered by license in LICENSE.proton file, so
there's no point in creating slightly different one inside a file.

dreamer commented:
Implemented additional tweaks feature: commandline parameters for specific binary. Also, introduced `PROTON_NO_TWEAKS` variable, so user can override all tweaks in one place and touched few lines of code in `proton.in` to make it more... python ;)

@aeikum  still waiting for *any* response regarding this PR

dreamer committed:
tweaks: Check fonts based on filename

dreamer committed:
tweaks: Workaround for RTCW

Based on feedback in #542, these environment variables help both Mesa
and NVIDIA users to run Return to Castle Wolfenstein (9010).

ihusky02 commented:
World of Tanks Blitz still doesn`t work with this tweaks?

@dreamer check message from stanislawtlolka@gmail.com

aeikum commented:
@dreamer It's a complicated question. We already have a mechanism for this kind of thing (the steam compat config variable), so we need to evaluate the pros and cons of doing it this way instead. I don't have an answer for you yet, sorry.

dreamer commented:
@aeikum The only problem I have with `STEAM_COMPAT_CONFIG` is lack of public access to the database - I can't browse it to see what fixes work for what titles, and (more importantly), I can't suggest fixes through PR or other processes (I guess Whitelist issue template exists for that purpose, maybe? But it covers only already existing `PROTON_*` variables).

I understand you don't want to complicate things and duplicate existing functionality - so I would be fine with dropping `TWEAKS_DB` variable and exposing specific tweaks through additional `PROTON_*` variables.

Another way forward could be for me to rename `PROTON_NO_TWEAKS` to `PROTON_COMMUNITY_TWEAKS` (disabled by default) - this way `tweaks.py` could serve as staging place for community PRs, before you adopt fixes into "steam compat config" mechanism.

plaes commented on tweaks.py:
```diff
@@ -60,12 +65,110 @@
+
+
+    def is_font_installed(self, file_name):  # pylint: disable=missing-docstring
+        return os.access(self.prefix + '/drive_c/windows/Fonts/' + file_name, os.F_OK)
```
os.path.join

dreamer replied:
I know `os.path.join` is an idiomatic way of creating paths in Python, but:

1) Rest of code uses string concatenation for this purpose
2) This code is not cross-platform, it is intended to run on Unix systems only
3) There are no user-supplied strings concatenated here - we can expect prefix and file_name to be properly escaped already (this is a shaky argument)

*I am not ruling out using os.path.join, merely explaining why I didn't use it for this PR.*

nanonyme replied:
I'd say the more interesting usage of os.path.join is avoiding double-path-separators. Forward slash works just fine on Windows even

dreamer replied:
@plaes @nanonyme - implemented, thank you for review :)

ihusky02 commented:
@dreamer something new, about run word of tanks blitz?

flibitijibibo commented:
Another possible use for this is the use of XnaToFna to fix games that depend on XNA to run; wine-mono doesn't cover C++/CLI assemblies so at the moment it's impossible to run these games without the .NET runtime. The way we have it set up, you target an exe file and it relinks the dependencies to use FNA instead, while also covering a couple formats we can't support (plain WMV/WMA files), and this removes the C++/CLI issue for most XNA games out there.

This is something I'm [looking into now](https://github.com/0x0ade/XnaToFna/issues/13), but the idea is that a really small zipfile can run XnaToFna on games known to depend on XNA (unfortunately the redists can be added as deps even if this isn't the case, so it'd be a manual process), then the game can run through Wine as usual.

Mind you, this doesn't cover every XNA game - only 4.0 is supported, and games using C++/CLI in any other capacity are out as well (old Steamworks C++/CLI wrappers, DirectInput wrappers, etc). The long-term solution would be to implement C++/CLI but that's a multi-year project for an _extremely_ small percentage of games out there.

This is a comically extreme example of how a game-specific tweak tool could be used, but it's worth a thought.

lucifertdark commented:
The game [586970] Hungry Flame needs Corefonts as well, other than that it works perfectly. I've added a report to SPCR just now but how long it'll take to show up, who knows?

dreamer committed:
tweaks: Turn on Glide emulation for Tomb Raider I

Override the DOSBox configuration distributed with the game by creating
a new *.conf file containing only settings, that need to be changed and
appending it to dosbox.exe arguments.

This game uses DOSBox patched with Glide support (something to consider
for #1566).  Unfortunately, Glide to OpenGL translation does not work
(leaving "Failed to Initialise OpenGL Window!" message in glide.err file
and nothing more), but Glide emulation works.

dreamer committed:
tweaks: Use os.path.join for path manipulation

dreamer committed:
tweaks: Add Core Fonts for Hungry Flame

dreamer commented:
Summary of changes:

* Rebase patches on top of current `proton_3.7` branch.
* Use rebase as an opportunity for some stylistic changes to tweaks.py (using `'` for strings consequently)
* Use os.path.join throughout tweaks.py
* Add functionality for creating .ini files in game directory and used it for creating override for dosbox configuration file for Tomb Raider I (224960)
* Add Hungry Flame (586970) to list of games needing Core Fonts installation

dreamer commented:
@flibitijibibo I tried to make Duck Game work with XnaToFna to possibly integrate it in tweaks, but had no success so far - I suspect a problem with my mono installation.

@kisak-valve @aeikum please take a look at dosbox fix specific for Tomb Raider in commit [ba97fe](https://github.com/ValveSoftware/Proton/pull/1423/commits/ba97feab99d17864201fdc0a95c09ee9db251e1a) - I don't think these kind of fixes are feasible with steam compat config mechanism. Did you have a chance to discuss internally if this way of distributing workarounds for games fits Proton?

HonkingGoose commented:
@dreamer you said to drop a comment if anybody finds a game with font issues? I have a game which has font issues for some people:

Could someone who understands fonts take a look at [this](https://github.com/ValveSoftware/Proton/issues/791#issuecomment-416818307)? A user has font related problems with Train Simulator (2018 version at the time of report). I don't have font related issues, but I use a standard Ubuntu 18.04 install.

I'm not entirely sure if this is relevant, but I see a lot of mentions with regards to fonts around here, so this seemed a good place to ask. If not feel free to tell me its not relevant here, and I'll delete the comment to keep things clear on this issue.

dreamer commented:
@HonkingGoose Existing corefonts workaround does not help in Train Simulator, so I can't include it here. I'll keep tabs on this problem, however - let's continue the discussion in #791.

====

I am closing this PR in favour #1745 (updating this one with rebased commits would create too much mess). Also, I believe Proton devs would prefer to stabilize 3.7 branch for games, that were already whitelisted so this PR is unlikely to be accepted.
